rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,275 @@
+
+```rust
+    // sending stream
+        let response = client.receive_stream(request).await?.into_inner();
+        println!("RESPONSE=\n{}", response.message);
+```
+
+Or if we want to go to bidirectional?
+
+```rust
+    // calling rpc
+        let mut response = client.bidirectional(request).await?.into_inner();
+    // listening on the response stream
+        while let Some(res) = response.message().await? {
+            println!("NOTE = {:?}", res);
+        }
+        Ok(())
+```
+
+This is obviously overly simplistic and a bit hand-wavey but, hopefully it expresses the value tonic provides in the form of these easy to consume abstractions. That seems to remain true if we want to do mutual TLS authentication or a few other more niche gRPC specific features which is fantastic. This library alone has me won over.
+
+The truth is that of the formats that have wide adoption, all of them have their pitfalls and while Protobufs bring with it some maintenance burden its still some of the _best_ tools with better integrations and libraries than most of the other things out there.
+
+As with protofbufs, HTTP/2 is pretty dang performant, it can provide users with some header compression, and it allows us to leverage client-side load-balancing in kubernetes. It feels like the best choice from that perspective.
+
+As an intentionally buried lede - I wrote this in the hopes that folks would share their opinions on the subject so you might be able to find some gaps in theorem here.
+
+## Prior Art
+
+Per the data format itself - I haven't found any single conglomerate conlusion about _any_ encoding format. However, existing projects using HTTP/2 and Protobufs (or gRPC over Tonic and Protobufs) are about a dime a dozen. These protocols and formats are used together in just about anything anymore.
+
+## Drawbacks
+
+The only drawback to adopting protobuf as the encoding format is that Protobufs can be slightly slower than other schema'd data formats and we don't shed the protobuf tooling overhead.
+
+Drawbacks to adopting HTTP/2 are that (like literally any other transport we could pick) it alters our kubernetes deployment handling and if we don't decide to continue to maintain other transports it means needing to appropriately handle the deprecation and removal of our TCP and HTTP/1.1 implementations fairly soon afterwards.
+
+## Alternatives
+
+One alternative to consider is that we just commit to providing multiple transports and stick with protobufs for encoding decoding. I think this approach should be considered thoroughly. Assuming we're using protobufs as the only encoding type, maintaining a few different transports might not be a huge maintenance overhead and it could give users some runtime flexibility that might add value.
```

Comment:
If I put on my Enterprise Ops hat there's an appeal to me in a product that exposes different transports but a common serialization format between them.